lib/figaro.rb

```diff
@@ -0,0 +1,63 @@
+require 'yaml'
+require 'erb'
+
+class Figaro
+  def self.env
+    @env ||= Environment.new
+  end
+
+  def self.setup(path)
+    env.setup(path)
+  end
+
+  def self.require_keys(keys)
+    env.require_keys(keys)
+  end
+
+  class Environment
+    def setup(path, env = Rails.env)
+      @config = {}
+      values = YAML.safe_load(File.read(path))
+
+      keys = Set.new(values.keys - %w[production development test])
+      keys |= values[env].keys
+
+      keys.each do |key|
+        value = values[env][key] || values[key]
+
+        ENV[key] = value if key == key.upcase
+
+        @config[key] = value
+      end
+
+      @config
+    end
+
+    def respond_to_missing?(method_name, _include_private = false)
+      key = method_name.to_s.tr('?!', '')
+      @config.key?(key)
+    end
+
+    def method_missing(method, *_args)
+      string_key = method.to_s
+      key = string_key.tr('?!', '')
```

Comment:
Do we care to continue supporting these Figaro-isms? It doesn't seem we're using the `?` support, and if we were, I don't think its behavior is correct here (returning a boolean). I see `!` is for [Figaro's lazy initialization](https://github.com/laserlemon/figaro#required-keys), but we only appear to use it in a handful of places and could maybe update those here?

Reply:
identity-idp/ $ rg 'Figaro\.env\.\w+!'
config/newrelic.yml
14:  license_key: <%= Figaro.env.newrelic_license_key! %>

config/database.yml
37:  database: <%= Figaro.env.database_name! %>

app/services/reports/base_report.rb
30:      "#{Figaro.env.s3_report_bucket_prefix!}.#{ec2_data.account_id}-#{ec2_data.region}"
47:      if Figaro.env.s3_reports_enabled! == 'false'

app/services/job_runner/runner.rb
25:      @disabled_jobs ||= JSON.parse(Figaro.env.recurring_jobs_disabled_names!)

Reply:
Ah, okay, I missed that `!` is directly related to required keys.

I was more wondering if we were expecting to preserve some of these existing behaviors.

If yes:

- `?`-suffixed environment variables should return a boolean
- [Handle env variable assignment? (`=`)](https://github.com/laserlemon/figaro/blob/6ae998e8209a08f91a165822abecffc0a0626814/lib/figaro/env.rb#L29)

If no:

- We could update existing references to `!` to just use the plain key
- We could remove this handling here

Reply:
The only goal of preserving existing behaviors is where removal would potentially break the app 🙂 

The `?` isn't used by us, so it wasn't included. The `ENV` assignment behavior was removed with the exception of cases like `RACK_TIMEOUT_SERVICE_TIMEOUT` mentioned at the top. Within our usage of Figaro, it was mostly circular where Figaro would set ENV from the YAML file, and then read from ENV later.

I removed the `!` behavior in https://github.com/18F/identity-idp/pull/4408/commits/687186d912ec043facd6e9122c51e77156a69aed and added those keys to the `require_keys` list.

Reply:
`ProductionDatabaseConfiguration` had some sneaky uses of the `!` suffix. It's a pretty narrow case that doesn't fit as a required key. It didn't feel like it was worth adding the `!` behavior back, so the class now checks and raises itself if a key is missing.